### ISSUE TITLE
ValidationFailure should be aware of it's RuleSet

### ISSUE THREAD

**Issue by ChaseFlorell** [FluentValidation/FluentValidation#1501]
**Is your feature request related to a problem? Please describe.**
When a `ValidationFailure` is constructed, it should be aware of the RuleSet name that it came from.

**Describe the solution you'd like**
Please add the following to `ValidationFailure`

```csharp
public string RuleSet { get; }
```

**Describe alternatives you've considered**
I've attempted to foreach all of my rulesets, but unfortunately I don't even see the validation failure when I do this.

```csharp
var failures = new Dictionary<string, ValidationResult>();

foreach (var ruleSet in _ruleSets)
{
    var result = await _validator.ValidateAsync(self, opts => opts.IncludeRuleSets(ruleSet));
    failures[ruleSet] = result; // unfortunately this is always IsValid = true;
    isValid = isValid && result.Errors.All(x => x.Severity != Severity.Error);
}
```

**Additional context**
Effectively I need to emit out validation failures based on the grouping of validation.
In other words, a particular segment of my overall object can be valid or invalid based on a RuleSet.

**Comment by JeremySkinner** [comment]
Hi, thanks for the suggestion.

The list of executed ruleset names is already present on the `ValidationResult`. Could you elaborate on the use case for needing it on the individual validation failures too?

**Comment by ChaseFlorell** [comment]
Effectively I need to know which RuleSet caused the ValidationFailure.

**Comment by JeremySkinner** [comment]
Could you help me understand why specifically you need to do that? What's your use case? Thanks!

**Comment by ChaseFlorell** [comment]
I'm doing real time validation of a custom Wizard in Xamarin Forms.

The wizard has multiple steps
Each step has a property CanMoveNext
The entire Wizard has a property Is valid

The user should be able to move next as long as their specific RuleSet is
valid, but not complete/finish until the entire validator Is valid.

**Comment by JeremySkinner** [comment]
Thanks, that makes sense. Yes that seems like a good solution, I'll get this added for a future version.

**Comment by JeremySkinner** [comment]
An individual rule can actually be part of multiple rulesets, so this will probably be an array of strings rather than a single string.

**Comment by ChaseFlorell** [comment]
What kind of effort would this look like to implement? Is there anything I can do to contribute or expedite?
________________________________

**Comment by JeremySkinner** [comment]
If you’d like to submit a pull request that’d be very welcome - I’ll happily review it. 

Off the top of my head It would involve the following:
- adding the new property to the ValidationFailure
- update the logic that creates the failure (inside PropertyValidator)  to set this, based on the RuleSets property of the parent rule 
- update the CustomValidator implementation to do this too (as it bypasses the standard failure creation process)
- add tests for the basic use case 
- add tests for when using Custom 
- add tests for checking that this is handled correctly when a rule is in multiple rulesets 
- every test should have 2 versions (to cover both sync and async workflow)

**Comment by JeremySkinner** [comment]
If you wanted a quicker solution that you could use now, you can call WithState in the rule to associate any arbitrary data with a failure (accessed via the failure’s CustomState property)

**Comment by ChaseFlorell** [comment]
I have achieved what I needed with Custom State. I still think this is a valid addition however.

**Comment by JeremySkinner** [comment]
I agree. If you’re interested in contributing the functionality then please feel free, otherwise I’ll try and look at it when I get some time

**Comment by jzabroski** [comment]
Really cool idea

**Comment by ChaseFlorell** [comment]
@JeremySkinner one thing I came across with CustomState is that it doesn't apply to the entire chain

```csharp
RuleFor(x => x.Email)
    .EmailAddress()
    .NotNull()
    .WithState(_ => "Email");
```

It would appear the CustomState is only applied to `NotNull` and not `EmailAddress`

So I guess this enhances my need to get this current feature built ¯\_(ツ)_/¯. I'll see if I can make some time.

**Comment by jzabroski** [comment]
@ChaseFlorell Can you workaround it with:

```c#
RuleFor(x => x.Email)
    .EmailAddress().WithState(_ => "Email")
    .NotNull().WithState(_ => "Email");
```

**Comment by JeremySkinner** [comment]
@ChaseFlorell Yes that's correct, custom state is always per-validator, it's not shared between the whole rule chain. However you can add an extension method that would apply it the whole chain quite easily:

```csharp
public static IRuleBuilderOptions<T, TProperty> RuleState<T, TProperty>(this IRuleBuilderOptions<T, TProperty> rule, Func<T, object> stateProvider) {
	var wrapper = new Func<PropertyValidatorContext, object>(ctx => stateProvider((T) ctx.InstanceToValidate));
	return rule.Configure(config => {
		foreach (var validator in config.Validators) {
			validator.Options.CustomStateProvider = wrapper;
		}
	});
}
```

Then you can call:

```csharp
RuleFor(x => x.Email).EmailAddress().NotNull().RuleState(_ => "Email");
```

It must be the last method in the chain, as it'll only apply to any validators that precede it.

**Comment by ChaseFlorell** [comment]
@jzabroski yes, that is the work around.

Some of our values have ~10 items in the chain, and some of our models have ~50 validatable properties, so it get's a little repetitive.

**Comment by jzabroski** [comment]
@ChaseFlorell Yeah, see Jeremy's reply above.  My point was if somehow CustomState was getting wiped out from one validation to the next, then that's a problem (but it's not).  It's easy to DRY up WithState via custom extension method.

**Comment by GuzzoLM** [comment]
Seems like nobody had the time to pick this up yet, so I'll give it a try.
